### ISSUE TITLE
Add a "focus on first tab" option to FocusTrap in CDK

### ISSUE THREAD

**Issue by gtranter** [angular/material2#10156]
#### Bug, feature request, or proposal:
Proposal

#### What is the expected behavior?
For dialogs and other containers such as drawers, menus etc. When the dialog or container is initialized to receive focus, instead of focusing the first tabbable element or the cdkFocusInitial directive's element, do not focus any element. Instead focus the first element upon the first tab or the last element upon shift-tab.

#### What is the current behavior?
Currently when using cdkFocusRegionStart and cdkFocusRegionEnd, the first tabbable element with the focus trap region receives initial focus. The cdkFocusInitial directive allows changing which component, but there is no mechanism to have no element initially focused. There are often use cases when no initial focus is a preferred behavior (e.g. dialogs where the primary action is non-reversible such as a delete operation).

#### What is the use-case or motivation for changing an existing behavior?
To support use cases where no initial focus is an important UX concern.

#### Is there anything else we should know?
This could be low effort. It appears that the wiring already in FocusTrap could easily support this. A new directive such as 'cdkFocusInitialTab' could be added to be the focus target on initial tab, and code changes made so that initial focus to the first tabbable element is not automatically applied when that directive is used. Usage could be something like:
```
<button mat-button cdkFocusRegionStart cdkFocusInitialTab>First</button>
...
<button mat-button cdFocusRegionEnd>Last</button>
```

**Comment by mischkl** [comment]
I also brought up similar concerns in https://github.com/angular/material2/issues/2879#issuecomment-371902277. As far as I can tell thi is what the `autoFocus` option is for.

```ts
MatDialog.open(MyDialog, {autoFocus: false});
```

However, this is of course only usable when calling `MatDialog.open()`. At the moment if you want to ensure that nothing is focused regardless of how the dialog is opened, you could add `tabIndex="-1"` to all of your buttons. I agree that a directive that can be added alongside `cdkFocusRegionStart` or `cdkTrapFocus` in order to disable autofocus would be nice.

Unfortunately I don't quite understand the choice of directive name `cdkFocusInitialTab` or how it should "be the focus target on initial tab" - isn't the focus target on initial tab *already* set when you specify the focus region??

If all the directive should do is to disable initial focus behavior, calling the directive something like `cdkFocusInitialNone` would make more sense to me.

**Comment by gtranter** [comment]
@mischkl You're right - autoFocus does turn off initial focus. But I think you've missed the point. This isn't strictly about initial focus, it is about **focus on first tab**. Turning off initial focus is just one part of it.

There are two things autoFocus doesn't address. First, autoFocus isn't configurable from a DOM template. Second - it doesn't solve the initial focus upon tab issue. Even if you use `cdkFocusInitial`, when you use `autoFocus: false` it ignores the component with `cdkFocusInitial` and upon first tab focuses the first component in the tab order. (Now maybe this is a bug and this should work - I don't know.)

Using tabIndex=-1 to avoid initial focus is bad practice in my opinion because it effectively destroys keyboard accessibility.

As for the name choice `cdkFocusInitialTab` it was just a suggestion, but your suggestion `cdkFocusInitialNone` fails to capture the intent of the feature, whereas `cdkFocusInitialTab` does. You are also right about the initial focus target being already set by the focus region - and that's exactly my point here - it is automatically set but there needs to be a way to specify whether a component should be automatically focused or focused only after the first tab press with no initial focus - thus `cdkFocusInitial` as well as `cdkFocusInitialTab`. `cdkFocusInitialTab' would turn off initial focus in addition to specifying which component receives focus on the first tab press.

**Comment by mischkl** [comment]
Thanks for the clarification, now I understand what you're trying to do! So here are my thoughts now that I understand.

First of all, the fact that the initial focusable element specified by `cdkFocusInitial` is ignored when you specify `autoFocus: false` sounds like a bug IMHO. If that were fixed the necessity for an additional directive would decrease to some extent.

However, in that case there would still be no way to disable autofocus via template for any use of `cdkFocusRegion*` or `cdkFocusTrap` (not just dialogs).

I propose that something like `cdkAutoFocus="false"` could be provided as a directive that could be added to same component as `cdkFocusRegionStart` or `cdkFocusTrap`. This could be used combination with the existing `cdkFocusInitial` in order to specify the first element to be focused, regardless of whether on tab or automatically.

This seems to me to more clearly communicate the intent of the directive (disabling autofocus) than something like `cdkFocusInitialTab`.

**Comment by gtranter** [comment]
> Just a heads up that we kicked off a community voting process for your feature request. There are 20 days until the voting process ends.

Only took 4 years for this - impressive.

**Comment by Maximaximum** [comment]
Totally ridiculous that this hasn't been addressed for 5 years already 🤦
